Fix issue where edge rails removes write_attribute_without_type_cast and implements default variants

geetfun opened the pull request:
This PR addresses the following two issues:

-  Edge Rails commit breaks Lockbox's use of #write_attribute_without_type_cast #101
- Rails 7 ActiveStorage introduces predefined variants which breaks Lockbox's variant handling #102

Both issues are related to upcoming changes in Rails 7.

Specifically:

- Rails removes #write_attribute_without_type_cast 
- Rails modifies ActiveStorage with default variants

I've added a rails70.gemfile in case you want to add it to your CI pipeline.

Let me know!

Thanks!

ankane commented on gemfiles/rails70.gemfile:
```diff
@@ -0,0 +1,16 @@
+gem "minitest", ">= 5"
+
+# This is the Rails 7.0.0.alpha commmit where Lockbox begins breaking
+gem 'rails', github: 'rails/rails', ref: '14ff158e73f17d2cb76f89f0d8c84a556f7589d9'
```
Let's test against `main` branch.

geetfun replied:
Sounds good, will make the edits.

ankane commented on lib/lockbox/model.rb:
```diff
@@ -192,10 +192,11 @@ def update_columns(attributes)
                 # same logic as Active Record
                 # (although this happens before saving)
                 attributes_to_set.each do |k, v|
-                  if respond_to?(:write_attribute_without_type_cast, true)
```
Let's keep the current logic to minimize any chance of breakage and add the new logic at the end.

geetfun replied:
Sounds good. I'll take a look at the code and will modify the logic so it finishes with the new logic at the end.

ankane commented on lib/lockbox/model.rb:
```diff
@@ -254,8 +255,13 @@ def reload
+                  attribute name, ActiveRecord::Type::Serialized.new(ActiveRecord::Type::String.new, attribute_type.coder)
+                end
+              elsif ActiveRecord::VERSION::STRING.to_f >= 6.1 && attributes_to_define_after_schema_loads[name.to_s].first.is_a?(Proc)
+                attribute_type = attributes_to_define_after_schema_loads[name.to_s].first.call(nil)
```
Let's keep the current logic (no `nil` argument).

ankane commented on lib/lockbox/model.rb:
```diff
@@ -381,10 +387,13 @@ def reload
 
                 # cache
                 # decrypt method does type casting
-                if respond_to?(:write_attribute_without_type_cast, true)
```
Same feedback as `update_columns` method.

ankane reviewed (commented):
Thanks for the PR @geetfun! Lockbox doesn't officially support Rails edge, but I'm down to try and keep things working if you'd like to submit PRs when things break. Just keep in mind that it's possible for Rails changes to introduce security issues. I left some feedback inline.

geetfun commented:
@ankane I've amended the PR:

- Rearranged the conditional logic in the two places as described in the review
- I've also updated the edge Rails Gemfile and renamed it to reflect that it's for Rails edge
- For the `call(nil)` thing, I've removed it from AR >= 6.1. For Rails edge, it requires the nil argument due to a change on the upstream Rails branch

Thanks for the review and let me know if you see anything else that needs modifying.

ankane commented:
Looks great, thanks @geetfun! (fwiw, I like the name `rails70` for the Gemfile, so going to restore that in a follow up commit)

geetfun commented:
Thanks again Andrew for the great gem!

ankane commented:
Hey @geetfun, just fyi, the `encrypts` method is now named `lockbox_encrypts` for Rails edge to prevent conflicts with Active Record's new `encrypts` method. It's [aliased in the Lockbox tests](https://github.com/ankane/lockbox/blob/16c350dd48ba8bb8cde180532e9a744f4ad9156d/test/support/active_record.rb#L5), but not in Rails edge apps.
